### PR TITLE
Add `Create` and `Edit and run` features for CustomRuns

### PR DESCRIPTION
**Pull request by jisoolee**
# Changes

The adding `Create` and `Edit and run` features for CustomRuns task and it is subdivided into 3 parts (will have 3 different commits per each part in one PR)
- [x] Part1. Adding Create and Edit and run buttons
- [x] Part2. Creating a Create page + Connecting the buttons to the proper Create/EditAndRun pages
- [x] Part3. Adding test cases

## Part1. Adding Create and Edit and run buttons
b1
b2
b3

## Part2. Creating a Create page + Connecting the buttons to the proper Create/EditAndRun pages
createcr

editandruncr

## Part3. Adding test cases
image

Connect to https://github.com/tektoncd/dashboard/issues/2437

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user facing changes (new features, significant UI changes, API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by AlanGreene** [comment]
Thanks @jisoolee. We need to make sure the Dashboard is always in a releasable state, so we'll either need to hide the buttons behind some kind of feature flag (query param or localStorage), or hold this PR until the create / edit page is merged first.

**Comment by jisoolee** [comment]
@AlanGreene Thank you for review and comment. I understand, yes I was also worrying about the release schedule (if I could not completed all the tasks before the next release). Please hold this PR until I open other PRs to complete the feature.

**Review by AlanGreene — CHANGES_REQUESTED** [review]
There's a lot of code in this related to the form-based UI which we are not adding for CustomRuns. Once that's removed there may be some opportunities to simplify / cleanup the remaining code.

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (jisoolee) [reply src/containers/CreateCustomRun/CreateCustomRun.jsx:472]: This part seems needed.
Without them, it shows when I tried to open Create page
image

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (jisoolee) [reply src/containers/CreateCustomRun/CreateCustomRun.jsx]: Thank you Alan. I updated the fields but could you double check if they are ok? Or do I need to put `spec.` in front of the names as you mentioned above? (Also do I need to remove some more except the above 4+5 fields? such as `labels`)

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [reply src/containers/CreateCustomRun/CreateCustomRun.jsx:472]: yes you must always return something from the render function
it looks like you have it right now, so it either returns the editor with the source run for a rerun, or the editor with the skeleton run for a new one 👍

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [reply src/containers/CreateCustomRun/CreateCustomRun.jsx]: `labels` is part of `metadata` so that's OK

File: src/containers/CustomRuns/CustomRuns.jsx
Comment (AlanGreene) [inline src/containers/CustomRuns/CustomRuns.jsx:334]: For now we'll always have at least 1 query param (`mode=yaml`) so we can simplify this and always construct the query string using URLSearchParams:

```suggestion
            const queryString = new URLSearchParams({
                ...(namespace !== ALL_NAMESPACES && { namespace }),
                // currently default is yaml mode
                mode: 'yaml'
              }).toString();
            }
            navigate(urls.customRuns.create() + `?${queryString}`);
```

File: src/containers/CustomRuns/CustomRuns.jsx
Comment (AlanGreene) [inline src/containers/CustomRuns/CustomRuns.jsx:253]: No need to change this string, we can continue using the existing one:

```suggestion
          id: 'dashboard.cancelTaskRun.actionText',
```

File: src/containers/CustomRuns/CustomRuns.jsx
Comment (AlanGreene) [inline src/containers/CustomRuns/CustomRuns.jsx:121]: These aren't used so they can be removed:

```suggestion
```

File: src/api/customRuns.js
Comment (AlanGreene) [inline src/api/customRuns.js:21]: This doesn't appear to be used

File: src/api/customRuns.js
Comment (AlanGreene) [inline src/api/customRuns.js:46]: not used

File: src/api/customRuns.js
Comment (AlanGreene) [inline src/api/customRuns.js:81]: As mentioned in [a previous review](https://github.com/tektoncd/dashboard/pull/3108#discussion_r1332042350), podTemplate and nodeSelector are not valid for a CustomRun so should be removed:
```suggestion
```

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:157]: There's still a lot of code in this component related to the form-based UI, for example, `checkFormValidation`, `handleAddLabel`, `handleNamespaceChange`, `handleSubmit`, etc.

These should all be removed. Then you can also remove any of the unused state, for example `initialParams`, `initialParamsState`, etc.

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:205]: Doesn't look like this is used

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:271]: These are all related to the form-based UI so are not needed.

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:339]: Not used.

There may be more. Please review any remaining functions to ensure they're actually called, or remove them if they're no longer needed.

**Review by AlanGreene — CHANGES_REQUESTED** [review]

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:23]: These aren't used and should be removed

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:27]: `KeyValueList` isn't used so the import should be removed
```suggestion
import { Loading } from '@tektoncd/dashboard-components';
```

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:38]: Not used

File: src/containers/CreateCustomRun/CreateCustomRun.test.jsx
Comment (AlanGreene) [inline src/containers/CreateCustomRun/CreateCustomRun.test.jsx:105]: Both of these are waiting for the same thing, you should be able to remove the second one.

Looks like this duplication is also in the CreatePipelineRun test, I've opened a PR to remove that one: https://github.com/tektoncd/dashboard/pull/3138

File: src/containers/CustomRuns/CustomRuns.jsx
Comment (AlanGreene) [inline src/containers/CustomRuns/CustomRuns.jsx:24]: not used

**Review by AlanGreene — CHANGES_REQUESTED** [review]
There's still quite a bit of unused code here that needs to be removed. I've added comments to the biggest pieces.

Most of this should have been caught by the linter so I'll have to debug that to figure out why it's not reporting errors.

For now you'll have to manually review each function, variable, and state field you've added to ensure they're actually used, and remove any that are not needed. When you remove a function, for example `handleSubmit`, check if there are additional functions or variables that can be cleaned up following its removal.

**Comment by AlanGreene** [comment]
For tracking purposes:

The issue with the linter not reporting issues was identified and resolved in https://github.com/tektoncd/dashboard/pull/3139

File: src/containers/CreateCustomRun/CreateCustomRun.test.jsx
Comment (briangleeson) [inline src/containers/CreateCustomRun/CreateCustomRun.test.jsx:104]: Is this test needed?

File: src/containers/CreateCustomRun/CreateCustomRun.jsx
Comment (briangleeson) [inline src/containers/CreateCustomRun/CreateCustomRun.jsx:64]:
```suggestion
  const [{ kind, labels, namespace, params }] = useState({
    ...initialState,
    customRef: '',
    kind: 'Custom',
    namespace: getNamespace()
  });
```

**Review by briangleeson — CHANGES_REQUESTED** [review]

**Review by briangleeson — APPROVED** [review]
/lgtm

**Pull request merged**
